IdentityCore/src/requests/MSIDNonceTokenRequest.m

```diff
@@ -0,0 +1,205 @@
+    }
+
+    if (_requestParameters.authority.metadata.tokenEndpoint)
+    {
+        [self executeNetworkRequestWithCompletion:completionBlock];
+        return;
+    }
+    
+    [_requestParameters.authority resolveAndValidate:YES
+                                   userPrincipalName:_requestParameters.accountIdentifier.displayableId
+                                             context:_requestParameters
+                                     completionBlock:^(NSURL __unused *openIdConfigurationEndpoint, BOOL __unused validated, NSError *error)
+    {
+        if (error)
+        {
+            completionBlock(nil, error);
+            return;
+        }
+        
+        [self->_requestParameters.authority loadOpenIdMetadataWithContext:self->_requestParameters
+                                                    completionBlock:^(__unused MSIDOpenIdProviderMetadata *metadata, NSError *openIdError)
+         {
+             
+             if (openIdError)
+             {
+                 completionBlock(nil, openIdError);
+                 return;
+             }
+             
+             [self executeNetworkRequestWithCompletion:completionBlock];
+         }];
+    }];
+}
+
+- (void)executeNetworkRequestWithCompletion:(nonnull MSIDNonceRequestCompletion)completionBlock
+{
+    MSIDHttpRequest *nonceRequest = [self configureNonceNetworkRequestForEndpoint:self.requestParameters.tokenEndpoint context:self.requestParameters];
+    [nonceRequest sendWithBlock:^(NSDictionary *response, NSError *error)
+    {
+        if (error)
+        {
+            if (completionBlock) completionBlock(nil, error);
+            return;
+        }
+        
+        if (![response isKindOfClass:[NSDictionary class]])
+        {
+            MSID_LOG_WITH_CTX(MSIDLogLevelError, self.requestParameters, @"Unexpected nonce response received");
+            NSError *nwError = MSIDCreateError(MSIDErrorDomain, MSIDErrorServerInvalidResponse, @"Unexpected nonce response", nil, nil, nil, nil, nil, YES);
+            if (completionBlock) completionBlock(nil, nwError);
+            return;
+        }
+        
+        NSString *nonce = [response msidStringObjectForKey:@"Nonce"];
+        
+        if ([NSString msidIsStringNilOrBlank:nonce])
+        {
+            MSID_LOG_WITH_CTX(MSIDLogLevelError, self.requestParameters, @"Didn't receive valid nonce in response");
+            NSError *nwError = MSIDCreateError(MSIDErrorDomain, MSIDErrorServerInvalidResponse, @"Didn't receive valid nonce in response", nil, nil, nil, nil, nil, YES);
+            if (completionBlock) completionBlock(nil, nwError);
+            return;
+        }
+        
+        [self.class cacheNonceForKey:self.requestParameters.authority.environment nonce:nonce];
+        if (completionBlock)
+        {
+            completionBlock(nonce, nil);
+        }
+    }];
+}
+
+- (MSIDHttpRequest *)configureNonceNetworkRequestForEndpoint:(NSURL *)endpoint context:(id<MSIDRequestContext>)context
```

Comment:
Can we move this into separate "network" request class in "network" layer?

Reply:
Separated into a different class.